pallets/crowdloans/src/lib.rs

```diff
@@ -959,6 +967,83 @@ pub mod pallet {
             }
             Ok(().into())
         }
+
+        #[pallet::weight(<T as Config>::WeightInfo::refund())]
+        #[transactional]
+        pub fn refund(origin: OriginFor<T>, crowdloan: ParaId) -> DispatchResult {
+            // 1. check phase, should be Closed or Failed
+            // 2. scan pending childstorage, mint RelayCurrency for contributor and kill contribution util reached RemoveKeysLimit.
+            // 3. if reached limit then return the function
+            // 4. if no and we already cleaned pending childstorage. Then we scan contributed childstorage and mint RelayCurrency and kill contribution for contributor util reached RemoveKeysLimit.
+            // If all contributions have been refunded then return AllRefunded event
+            // if not then return PartiallyRefunded event
+            Ok(())
+        }
+
+        #[pallet::weight(<T as Config>::WeightInfo::dissolve_vault())]
+        #[transactional]
+        pub fn dissolve_vault(
+            origin: OriginFor<T>,
+            crowdloan: ParaId,
+            lease_start: LeasePeriod,
+            lease_end: LeasePeriod,
+        ) -> DispatchResult {
+            // users who can create vaults can dissolve them
+            T::CreateVaultOrigin::ensure_origin(origin)?;
+
+            // 1. check phase, should be Closed or Failed or Expired
+            let vault = Self::vaults((&crowdloan, &lease_start, &lease_end))
+                .ok_or(Error::<T>::VaultDoesNotExist)?;
+
+            ensure!(
+                vault.phase == VaultPhase::Closed
+                    || vault.phase == VaultPhase::Failed
+                    || vault.phase == VaultPhase::Expired,
+                Error::<T>::IncorrectVaultPhase
+            );
+
+            // 2. check flying, pending, contributed childstorage, should have 0 contributions and `vault.contributed + vault.flying + vault.pending == 0` otherwise return back `NotReadyToDissolve`
+            let total_completed_contributions =
+                Self::contribution_iterator(vault.trie_index, ChildStorageKind::Contributed)
+                    .fold(0u128, |sum, (_account, (amount, _ref_code))| sum + amount);
+
+            let total_pending_contributions =
+                Self::contribution_iterator(vault.trie_index, ChildStorageKind::Pending)
+                    .fold(0u128, |sum, (_account, (amount, _ref_code))| sum + amount);
+
+            let total_flying_contributions =
+                Self::contribution_iterator(vault.trie_index, ChildStorageKind::Flying)
+                    .fold(0u128, |sum, (_account, (amount, _ref_code))| sum + amount);
+
+            ensure!(
+                total_completed_contributions
+                    + total_flying_contributions
+                    + total_pending_contributions
+                    == 0,
+                Error::<T>::NotReadyToDissolve
+            );
+
+            ensure!(
+                vault.contributed + vault.flying + vault.pending == 0,
```

Comment:
we can refactor the `total_contribution` to be the sum of `vault.contributed + vault.flying + vault.pending`

Reply:
and we can add `total_contribution_in_childstorage` to be the sum of the three childstorages

Reply:
also we can maybe have a helper function which sum a childstorage